gpt_engineer/ai.py

```diff
@@ -23,72 +39,134 @@ class TokenUsage:
 
 
 class AI:
-    def __init__(self, model="gpt-4", temperature=0.1):
+    def __init__(self, model_dir=None, model_name="gpt-4", temperature=0.1):
         self.temperature = temperature
-        self.model = model
+        if model_dir:
+            self.model_dir = Path(model_dir)
+        else:
+            # __file__ is "<package_dir>/gpt-engineer/ai.py"
+            # therefore .parent.parent/models is "<package_dir>/models"
+            self.model_dir = Path(__file__).resolve().parent.parent / "models"
+        self.model_name = fallback_model(model_name)
+        self.llm = None
+
+        llm_filename = self.model_dir / f"{model_name}.yaml"
+        try:
+            logging.info(f"LLM file name: {llm_filename}")
+            self.llm = load_llm(llm_filename)
+        except Exception as e:
+            raise RuntimeError(
+                f"Unable to load LLM {model_name} from file {llm_filename}", e
+            )
 
         # initialize token usage log
         self.cumulative_prompt_tokens = 0
         self.cumulative_completion_tokens = 0
         self.cumulative_total_tokens = 0
         self.token_usage_log = []
 
+        # ToDo: Adapt to arbitrary model, currently assumes model using tiktoken
```

Comment:
Unfortunately, LangChain doesn't give access to its models' tokenizers. So we would either (i) wait until LC does so, or (ii) implement each tokenizer for each model into gpt-engineer. Imo (ii) doesn't make sense because we want to let LC handle all the interactions with models, and focus on coding-specific stuff.